apps/push_originserver_multithread_linux.c

```diff
@@ -0,0 +1,373 @@
+{
+	oc_rep_open_object(root, rep);
+	oc_rep_set_int(rep, power, power);
+	oc_rep_set_int(rep, brightness, brightness);
+	oc_rep_close_object(root, rep);
+}
+
+static void register_resources(void)
+{
+  /* light2 is not pushable */
+	res2 = oc_new_resource(NULL, resource_uri2, 1, 0);
+	oc_resource_bind_resource_type(res2, resource_rt);
+	oc_resource_bind_resource_interface(res2, OC_IF_RW);
+	oc_resource_set_default_interface(res2, OC_IF_RW);
+	oc_resource_set_discoverable(res2, discoverable);
+	oc_resource_set_observable(res2, observable);
+	oc_resource_set_request_handler(res2, OC_GET, cb_retrieve_light2, NULL);
+	oc_add_resource(res2);
+
+	/* light1 is pushable */
+	res = oc_new_resource(NULL, resource_uri, 1, 0);
+	oc_resource_bind_resource_type(res, resource_rt);
+	oc_resource_bind_resource_interface(res, OC_IF_RW);
+	oc_resource_set_default_interface(res, OC_IF_RW);
+	oc_resource_set_discoverable(res, discoverable);
+	oc_resource_set_observable(res, observable);
+	/* make a Resource pushable */
+	oc_resource_set_pushable(res, pushable);
+	oc_resource_set_request_handler(res, OC_GET, cb_retrieve_light1, NULL);
+	/* set PUSH payload builder */
+	res->payload_builder = build_light_payload;
+	oc_add_resource(res);
+}
+
+static void signal_event_loop(void)
+{
+	pthread_mutex_lock(&mutex);
+	pthread_cond_signal(&cv);
+	pthread_mutex_unlock(&mutex);
+}
+
+void handle_signal(int signal)
+{
+	(void) signal;
+	signal_event_loop();
+	quit = 1;
+}
+
+static void*
+process_func(void *data)
+{
+	(void) data;
+	oc_clock_time_t next_event;
+
+	while (quit != 1)
+	{
+		pthread_mutex_lock(&app_mutex);
+		next_event = oc_main_poll();
+		pthread_mutex_unlock(&app_mutex);
+		pthread_mutex_lock(&mutex);
+		if (next_event == 0)
+		{
+			pthread_cond_wait(&cv, &mutex);
+		}
+		else
+		{
+			ts.tv_sec = (next_event / OC_CLOCK_SECOND);
+			ts.tv_nsec = (next_event % OC_CLOCK_SECOND) * 1.e09 / OC_CLOCK_SECOND;
+			pthread_cond_timedwait(&cv, &mutex, &ts);
+		}
+		pthread_mutex_unlock(&mutex);
+	}
+
+	pthread_exit(0);
+}
+
+void print_menu(void)
+{
+	pthread_mutex_lock(&app_mutex);
+	printf("=============== Light A =============\n");
+	printf("1. Change brightness(%d) of light #1\n", brightness);
+	printf("2. Change power(%d) of light #1\n", power);
+	printf("3. Change brightness(%d) of light #2\n", brightness2);
+	printf("4. Change power(%d) of light #2\n", power2);
+	printf("0. Quit\n");
+	printf("=====================================\n");
+	pthread_mutex_unlock(&app_mutex);
+}
+
+int main(void)
+{
+	int init = 0;
+	struct sigaction sa;
+	sigfillset(&sa.sa_mask);
+	sa.sa_flags = 0;
+	sa.sa_handler = handle_signal;
+	sigaction(SIGINT, &sa, NULL);
+
+	static const oc_handler_t handler =
+	{ .init = app_init, .signal_event_loop = signal_event_loop,
+			.register_resources = register_resources };
+
+#ifdef OC_STORAGE
+	oc_storage_config("./push_originserver_multithread_linux_creds");
+#endif /* OC_STORAGE */
+
+	if (pthread_mutex_init(&mutex, NULL) < 0)
```

Comment:
`pthread_mutex_init(&mutex, NULL) != 0`

Fixed in master already. https://linux.die.net/man/3/pthread_mutex_init - `pthread_mutex_init` returns 0 on success and error number (which are >0) on failure. (Same with `pthread_cond_init`.)

Reply:
done